Update the txome dimensionality reduction module now that files are split (#137)
AlexsLemonade/OpenPBTA-analysis: analyses/transcriptomic-dimension-reduction/README.md

# Dimension reduction: gene expression data

This module runs dimension reduction—namely PCA, UMAP, and t-SNE—on gene expression data from kallisto and RSEM.

> To the run the whole pipeline and produce 4 plots with panels for all three approaches

Reviewer: What are the three approaches being referred to? This itemized list doesn't make it clear to me.

Author: PCA, UMAP, t-SNE

1. RSEM stranded data
2. RSEM poly-A data
3. kallisto stranded data
4. kallisto poly-A data

> with points colored by broad histology, simply run:

Reviewer: ```suggestion
To run the whole pipeline and produce 4 plots with panels for all three approaches with points colored by broad histology, simply run:
```

```
./dimension-reduction-plots.sh
```

> ## Notes on further applications

Reviewer: ```suggestion
## Notes on further applications
```
Seems like this is the whole README, so feels weird to call it "Notes" maybe just *General Usage* or something?

To document this module beyond what is in the individual files, we list some example applications or extensions below.

Running dimension reduction, creating lists of plots, and generating the multipanel plots can all be run separately using the appropriate Rscripts in `scripts/`:

```
generate-multipanel-plot.R
get-plot-list.R
run-dimension-reduction.R
```

### Repeat dimension reduction step with different seeds

##### Command line

If we wanted to run UMAP multiple times with different seed values on the kallisto poly-A data, but _skip t-SNE_, we could do the following:

```sh
seeds=(309 4642 2019 2994 5783)

for seed in "${seeds[@]}"
do
  Rscript --vanilla scripts/run-dimension-reduction.R \
    --expression ../../data/pbta-gene-expression-kallisto.polya.rds \
    --metadata ../../data/pbta-histologies.tsv \
    --filename_lead kallisto_polyA_${seed} \
    --output_directory results \
    --seed ${seed} \
    --skip_tsne
done

```

> Unfortunately, this approach will also run PCA 5 times. 

Reviewer: Should it have an option to not do this?

Author: I think that use case is covered in the R section below. The way to do this would be to include logic to skip each of the PCA, UMAP, and t-SNE (or enable each step depending on how you did it), which is easy enough to do if it looks like it will come up often. I think a more likely scenario is that someone picked one of the three methods for display and is running it multiple times or tweaking the plots.

##### R

It's also possible to repeat this UMAP step 5 times within R:

> ```R

Reviewer: I think technically most markdowns want this to be a lower case `r` but I am unsure if this matters for anything.

Author: It does not, I just checked.

# load the required custom functions
source("util/dimension-reduction-functions.R")

# read in kallisto poly-A data and transpose

> expression_file <- "../../data/pbta-gene-expression-kallisto.polya.rds"

Reviewer: Should you use/encourage the use of `file.path`?

Author: I think in this setting this is easier to _read_. That was my rationale.

> # Read in expression data

Reviewer: Sometimes you capitalize your comments and sometimes you don't. This doesn't matter but I'm just stating this because it's something my brain notices.

Author: I think technically @cbethell capitalizes comments and I don't 😁

                              method = "UMAP",
                              seed = seed,
                              model_filename = paste("kallisto_polyA", seed, "UMAP.RDS"),
                              output_directory = "results",
                              neighbors_parameter = 15)
}
```

### Generate plots colored by short histology

To add multipanel plots with points colored by `short_histology` (must be a column in `pbta-histologies.tsv`), we can run the following:

```sh
# Change the point color variable when
COLOR=short_histology bash 02-get-dimension-reduction-plot-lists.sh
# This will make plots for anything with data
bash 03-multipanel-plots.sh
```

This will plot the first two variables for each approach.
Currently, this does not matter for the way t-SNE and UMAP are implemented.
The PCA plots are limited to PC1 and PC2.

### Plot additional principal components

If we wanted to plot additional PCs for the RSEM stranded data in R, we can take advantage of the custom `plot_dimension_reduction` function and use the following R code:

```R
# load the required custom functions
source("util/dimension-reduction-functions.R")

> # read in the data.frame that has the PCA info and the metadata

Reviewer: Could use some free lines between these comments and the code.

pca_df <- readr::read_tsv("results/rsem_stranded_pca_scores_aligned.tsv")
# plot itself - PC7 and PC10
plot_dimension_reduction(aligned_scores_df = pca_df,
                         point_color = "broad_histology",
                         x_label = "PC7",
                         y_label = "PC10",
                         score1 = 7,
                         score2 = 10)
```

> The `point_color` argument of `plot_dimension_reduction` can be changed to any variable that is in the TSV file read in.

Reviewer: ```suggestion
The `point_color` argument of `plot_dimension_reduction` can be changed to any variable that is a column in the original TSV file.
```
